README.md
# yandex_sign

Flutter plugin for yandex map sign and yandex navigator sign.

__Disclaimer__: This project uses Yandex sign algorithm which discribed 
in [commercial_use_signature_navigator](https://yandex.ru/dev/yandex-apps-launch/navigator/doc/concepts/navigator-commercial-use-signature-docpage/)
and [commercial_use_signature_yandexmaps](https://yandex.ru/dev/yandex-apps-launch/maps/doc/concepts/yandexmaps-commercial-use-signature-docpage/)

## Getting Started

### Generate your API Key

Get key and client Id from [yandex_navigator](https://yandex.ru/dev/yandex-apps-launch/navigator/doc/concepts/navigator-commercial-use-signature-docpage/) or [yandexmaps](https://yandex.ru/dev/yandex-apps-launch/maps/doc/concepts/yandexmaps-commercial-use-signature-docpage/)

### Initilazing for IOS
1. Convert key.pem file to key.der file with openssl `$ openssl rsa -in key.pem -out key.der -outform DER`
2. Add file key.der to ios project Bundle [example](https://stackoverflow.com/questions/10247680/adding-resource-files-to-xcode)


### Initilazing for Android
1. Get String(base64) from file key.pem
2. Add to local variable for flutter project like in example

### Usage

Example:

```dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:yandex_sign/yandex_sign.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _signUrl = 'Unknown';


  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  Future<void> initPlatformState() async {
    String signUrl;
    try {
      final url = 'yandexnavi://build_route_on_map?lat_to=55.70500&lon_to=37.587200&client=007';
      final androidKey = '''-----BEGIN RSA PRIVATE KEY-----
MIIBOgIBAAJBALUwoxjKUZtMiFPe1K4XAPDYZCX1b6wOM8UXeGS5EbsJe/h12VM1
asx9rtVXJ1Kwhb1N7KcSFc1l9wl2gqe1f8cCAwEAAQJAGV0Dl7bKkai29LaeovgJ
Y62G52FiDn22dYKBkefuOXYwFBSD6+pbHby3m5oBEy51aZJ+QQBIv+RjHN1h1aUQ
8QIhAN9hr9dxgzKYw8FUQBZ4qgd6hRSjb/ryfjgiF5qPt2FbAiEAz6XFDvwkbwE5
FLrO8MigLVXuEuhiCsbkwn8aobBUGwUCIBGlTNeu/tcrXCNUfW+I/p1ynzqfIoRn
TXMvtj+eZLULAiB/n4s2YpKiB0ZmD0sRgr2wH5hr1pgrt4LyZ9yedBm9YQIhAMIB
oRxLAAcDDIlC6vDrYk94nCT0OQ6ZH2YK3rxu8iyy
-----END RSA PRIVATE KEY-----
''';
      final iosNameOfDerFile = 'private_key';
      //Sign url for both android & ios with key
      signUrl = await YandexSign.getSignUrl(url, androidKey, iosNameOfDerFile);
    } on PlatformException {
      signUrl = 'Failed to get platform version.';
    }
    
    if (!mounted) return;

    setState(() {
      _signUrl = signUrl;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Text('Running on: $_signUrl\n'),
        ),
      ),
    );
  }
}
```

For more info see example

### Features

- [X] iOS Support
- [X] Android Support


CHANGELOG.md
## 0.9.2

* reformat change log
* add readme

## 0.9.1

* fix the Android Gradle plugin supports only Kotlin Gradle plugin version 1.3.10 and higher. Project 'yandex_sign' is using version 1.2.75.
* fix the bug with overriding android minSdkVersion in AndroidManifest.xml in android packege of plugin

## 0.9.0

* Add plugin to github and publish.


test/yandex_sign_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const MethodChannel channel = MethodChannel('yandex_sign');
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:yandex_sign_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


ios/Classes/SwiftYandexSignPlugin.swift
import Flutter
import UIKit
import Security
import CommonCrypto

@available(iOS 10.0, *)
public class SwiftYandexSignPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "yandex_sign", binaryMessenger: registrar.messenger())
    let instance = SwiftYandexSignPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  @available(iOS 10.0, *)
  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard call.method == "getSignUrl" else {
            result(FlutterMethodNotImplemented)
            return
        }
    self.getSignUrl(result: result, arguments: call.arguments)
  }

  @available(iOS 10.0, *)
  private func getSignUrl(result: FlutterResult, arguments: Any?) {
          let params = arguments as! [String: Any]
          let url = params["url"] as! String
          let iosNameOfDerFile = params["iosNameOfDerFile"] as! String

          let signature = signString(string: url, key: loadCert(iosNameOfDerFile: iosNameOfDerFile))
          result(signature.addingPercentEncoding(withAllowedCharacters: .alphanumerics));
      }
      @available(iOS 10.0, *)
      // Зашифровывает SHA256 с помощью DER-ключа.
      func signString(string: String, key: SecKey) -> String {
          let messageData = string.data(using:String.Encoding.utf8)!
          var hash = Data(count: Int(CC_SHA256_DIGEST_LENGTH))

          _ = hash.withUnsafeMutableBytes {digestBytes in
              messageData.withUnsafeBytes {messageBytes in
                  CC_SHA256(messageBytes, CC_LONG(messageData.count), digestBytes)
              }
          }

          let signature = SecKeyCreateSignature(key,
                                                SecKeyAlgorithm.rsaSignatureDigestPKCS1v15SHA256,
                                                hash as CFData,
                                                nil) as Data?

          return (signature?.base64EncodedString())!
      }

      @available(iOS 10.0, *)
      func loadCert(iosNameOfDerFile: String) -> SecKey {
          let certificateData = NSData(
              contentsOf:Bundle.main.url(forResource: iosNameOfDerFile, withExtension: "der")!
          )
          let options: [String: Any] =
              [kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
               kSecAttrKeyClass as String: kSecAttrKeyClassPrivate,
               kSecAttrKeySizeInBits as String: 512]

          let key = SecKeyCreateWithData(certificateData!,
                                         options as CFDictionary,
                                         nil)

          return key!
      }
}


android/src/main/kotlin/dsishokov/yandex_sign/YandexSignPlugin.kt
package dsishokov.yandex_sign

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar
import android.util.Base64
import android.util.Log
import java.net.URLEncoder
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.EncodedKeySpec
import java.security.KeyFactory
import java.security.Signature

class YandexSignPlugin: MethodCallHandler {
  companion object {
    @JvmStatic
    fun registerWith(registrar: Registrar) {
      val channel = MethodChannel(registrar.messenger(), "yandex_sign")
      channel.setMethodCallHandler(YandexSignPlugin())
    }
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    if (call.method == "getSignUrl") {
      result.success(getSignUrl(call.arguments))
    } else {
      result.notImplemented()
    }
  }

  private fun getSignUrl(arguments: Any): String {
    val params = arguments as Map<String, *>
    val url = params["url"] as String
    val key = params["androidKey"] as String
    val data = sha256rsa(key, url)

    return data
  }

  // Формирует подпись с помощью ключа.
  fun sha256rsa(key: String, data: String): String {
    val trimmedKey = key.replace("-----BEGIN RSA PRIVATE KEY-----", "")
            .replace("-----END RSA PRIVATE KEY-----", "")
            .replace("\\s".toRegex(), "")
    Log.d("KOTLIN", trimmedKey)
    try {
      val result = Base64.decode(trimmedKey, Base64.DEFAULT)
      val factory = KeyFactory.getInstance("RSA")
      val keySpec = PKCS8EncodedKeySpec(result)
      val signature = Signature.getInstance("SHA256withRSA")
      signature.initSign(factory.generatePrivate(keySpec))
      signature.update(data.toByteArray())

      val encrypted = signature.sign()
      return URLEncoder.encode(Base64.encodeToString(encrypted, Base64.NO_WRAP), "UTF-8")
    } catch (e: Exception) {
      throw SecurityException("Error calculating cipher data. SIC!")
    }

  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:yandex_sign/yandex_sign.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _signUrl = 'Unknown';


  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    String signUrl;
    // Platform messages may fail, so we use a try/catch PlatformException.
    try {
      final url = 'yandexnavi://build_route_on_map?lat_to=55.70500&lon_to=37.587200&client=007';
      final androidKey = '''-----BEGIN RSA PRIVATE KEY-----
MIIBOgIBAAJBALUwoxjKUZtMiFPe1K4XAPDYZCX1b6wOM8UXeGS5EbsJe/h12VM1
asx9rtVXJ1Kwhb1N7KcSFc1l9wl2gqe1f8cCAwEAAQJAGV0Dl7bKkai29LaeovgJ
Y62G52FiDn22dYKBkefuOXYwFBSD6+pbHby3m5oBEy51aZJ+QQBIv+RjHN1h1aUQ
8QIhAN9hr9dxgzKYw8FUQBZ4qgd6hRSjb/ryfjgiF5qPt2FbAiEAz6XFDvwkbwE5
FLrO8MigLVXuEuhiCsbkwn8aobBUGwUCIBGlTNeu/tcrXCNUfW+I/p1ynzqfIoRn
TXMvtj+eZLULAiB/n4s2YpKiB0ZmD0sRgr2wH5hr1pgrt4LyZ9yedBm9YQIhAMIB
oRxLAAcDDIlC6vDrYk94nCT0OQ6ZH2YK3rxu8iyy
-----END RSA PRIVATE KEY-----
''';
      final iosNameOfDerFile = 'private_key';
      signUrl = await YandexSign.getSignUrl(url, androidKey, iosNameOfDerFile);
    } on PlatformException {
      signUrl = 'Failed to get platform version.';
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _signUrl = signUrl;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Text('Running on: $_signUrl\n'),
        ),
      ),
    );
  }
}
